pkg/kubectl/cmd/exec/exec.go

```diff
@@ -266,11 +294,34 @@ func (o *StreamOptions) SetupTTY() term.TTY {
 
 // Run executes a validated remote execution against a pod.
 func (p *ExecOptions) Run() error {
-	pod, err := p.PodClient.Pods(p.Namespace).Get(p.PodName, metav1.GetOptions{})
-	if err != nil {
-		return err
+	var err error
+	// we still need legacy pod getter when PodName in ExecOptions struct is provided,
+	// since there are any other command run this function by providing Podname with PodsGetter
+	// and without resource builder, eg: `kubectl cp`.
+	if len(p.PodName) != 0 {
+		p.Pod, err = p.PodClient.Pods(p.Namespace).Get(p.PodName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		builder := p.Builder().
+			WithScheme(scheme.Scheme, scheme.Scheme.PrioritizedVersionsAllGroups()...).
+			NamespaceParam(p.Namespace).DefaultNamespace().
+			ResourceNames("pods", p.ResourceName)
```

Comment:
This is wrong, you need to pass the actual type and name here, compare with https://github.com/kubernetes/kubernetes/blob/fd98e7abf042251bb094a5535d67f53d36021246/pkg/kubectl/cmd/attach/attach.go#L230. Also I'd encourage you to switch from this conditions to a code similar in the linked attach.

Reply:
It's not wrong, it's the right way to handle single arguments while maintaining backwards compatibility with the existing format. With your suggestion, exec will now support spaces between resource types and names.

Reply:
Yes, if I apply something like this.

```
builder := p.Builder().
	WithScheme(scheme.Scheme, scheme.Scheme.PrioritizedVersionsAllGroups()...).
	NamespaceParam(p.Namespace).DefaultNamespace()

if len(p.PodName) != 0 {
	builder.ResourceNames("pods", p.PodName)
} else {
	builder.ResourceNames("pods", p.ResourceName)
}
```

it's cause a panic on `kubectl cp` command. because in `kubectl cp` does not fill the `Builder` in `ExecOptions` struct https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cp/cp.go#L506-L526

Reply:
I think what @soltysh meant is to end up with is something like the following:
```go
	if len(p.PodName) != 0 {
		p.Pod, err = p.PodClient.Pods(p.Namespace).Get(p.PodName, metav1.GetOptions{})
		if err != nil {
			return err
		}
	} else {
		b := p.Builder().
			WithScheme(scheme.Scheme, scheme.Scheme.PrioritizedVersionsAllGroups()...).
			NamespaceParam(p.Namespace).DefaultNamespace()

		switch len(o.Resources) {
		case 1:
			b.ResourceNames("pods", o.Resources[0])
		case 2:
			b.ResourceNames(o.Resources[0], o.Resources[1])
}
,,,

Reply:
hmmm, I misunderstood that 😄 

so like @kargakis say, whether `kubectl exec` will now support spaces between resource types and names?

Reply:
I think it's an acceptable thing to support but to be fair the original issue was about adding `type/name` support. I'd prefer to get this merged, and add `type name` support in a follow-up (also needs to be added in other commands such as `kubectl logs`)

Reply:
if kubectl exec support spaces between resource types and names like kubectl attach. how we can mapping between kubectl exec arguments and command with argument? like this `POD COMMAND [ARG]`, `TYPE/NAME COMMAND [ARG]` and `TYPE NAME COMMAND` :sweat_smile:

Reply:
I'd rather get it fixed in this PR while this is being modified. I know from experience that followups might slip away and that won't happen, so while at it please straighten this one up.

Reply:
@soltysh maybe @prksu's comment wasn't clear enough? Your suggestion about supporting spaces between resource types and names is actually a breaking one in `kubectl exec`.

Currently the following `kubectl exec` invocation is valid:
```console
$ kubectl exec weave-net-6xblq date
Defaulting container name to weave.
Use 'kubectl describe pod/weave-net-6xblq -n kube-system' to see all of the containers in this pod.
Mon Apr  8 20:30:45 UTC 2019
```
`date` is executed in the container (without having to pass `--` in order to separate between the kubernetes resource and the command). With your suggestion `weave-net-6xblq` would be parsed as a resource type and `date` would be parsed as a resource name, right?

Reply:
@kargakis maybe what @soltysh wants is just passing the actual type and name in resource builder. without supporting spaces between resource types and names. it's right @soltysh ?

Reply:
i just updated. PTAL